File: day 23/AOC23.py
# from collections import defaultdict

# # Step 1: Read connections from file
# with open('Lan.txt', 'r') as file:
#     connections = [line.strip() for line in file if line.strip()]

# # Step 2: Build the graph as an adjacency list
# graph = defaultdict(set)
# for connection in connections:
#     a, b = connection.split("-")
#     graph[a].add(b)
#     graph[b].add(a)

# # Step 3: Find all triangles
# triangles = set()
# for node in graph:
#     neighbors = graph[node]
#     for neighbor in neighbors:
#         # Find common neighbors between `node` and `neighbor`
#         common_neighbors = neighbors & graph[neighbor]
#         for common in common_neighbors:
#             # A triangle is formed by (node, neighbor, common)
#             triangle = tuple(sorted([node, neighbor, common]))
#             triangles.add(triangle)

# # Step 4: Filter triangles where at least one name starts with 't'
# filtered_triangles = [triangle for triangle in triangles if any(comp[0] == 't' for comp in triangle)]

# # Step 5: Output the result
# print("Total triangles with at least one computer starting with 't':", len(filtered_triangles))
# print("Triangles:", filtered_triangles)


#############################################################

from collections import defaultdict

# Step 1: Read connections from file
with open('Lan.txt', 'r') as file:
    connections = [line.strip() for line in file if line.strip()]

# Step 2: Build the graph as an adjacency list
graph = defaultdict(set)
for connection in connections:
    a, b = connection.split("-")
    graph[a].add(b)
    graph[b].add(a)

# Step 3: Find the largest clique (fully connected set)
def find_largest_clique(graph):
    def is_clique(nodes):
        return all(b in graph[a] for a in nodes for b in nodes if a != b)

    largest_clique = set()
    nodes = list(graph.keys())

    # Check all subsets of nodes for being a clique
    from itertools import combinations
    for size in range(len(nodes), 0, -1):
        for subset in combinations(nodes, size):
            if is_clique(subset):
                return set(subset)
    return largest_clique

largest_clique = find_largest_clique(graph)

# Step 4: Generate the password
password = ",".join(sorted(largest_clique))

# Step 5: Output the result
print("Largest clique size:", len(largest_clique))
print("Password to the LAN party:", password)

File: day 23/AOC23.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strings"
)

func bronKerbosch(r, p, x map[string]bool, graph map[string]map[string]bool, cliques *[][]string) {
	if len(p) == 0 && len(x) == 0 {
		var clique []string
		for node := range r {
			clique = append(clique, node)
		}
		*cliques = append(*cliques, clique)
		return
	}

	pCopy := make(map[string]bool)
	for k, v := range p {
		pCopy[k] = v
	}

	for node := range pCopy {
		newR := make(map[string]bool)
		for k, v := range r {
			newR[k] = v
		}
		newR[node] = true

		newP := make(map[string]bool)
		newX := make(map[string]bool)

		for neighbor := range graph[node] {
			if p[neighbor] {
				newP[neighbor] = true
			}
			if x[neighbor] {
				newX[neighbor] = true
			}
		}

		bronKerbosch(newR, newP, newX, graph, cliques)

		delete(p, node)
		x[node] = true
	}
}

func main() {
	file, err := os.Open("Lan.txt")
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		return
	}
	defer file.Close()

	graph := make(map[string]map[string]bool)
	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		parts := strings.Split(scanner.Text(), "-")
		if len(parts) == 2 {
			a, b := parts[0], parts[1]
			if graph[a] == nil {
				graph[a] = make(map[string]bool)
			}
			if graph[b] == nil {
				graph[b] = make(map[string]bool)
			}
			graph[a][b] = true
			graph[b][a] = true
		}
	}

	p := make(map[string]bool)
	for node := range graph {
		p[node] = true
	}

	var cliques [][]string
	bronKerbosch(make(map[string]bool), p, make(map[string]bool), graph, &cliques)

	// Find largest clique
	var largest []string
	for _, clique := range cliques {
		if len(clique) > len(largest) {
			largest = clique
		}
	}

	sort.Strings(largest)
	password := strings.Join(largest, ",")

	fmt.Printf("Largest clique size: %d\n", len(largest))
	fmt.Printf("Password to the LAN party: %s\n", password)
}
